Create EIP-4804 (#4995)
ethereum/EIPs: EIPS/eip-4804.md

---
eip: 4804
title: Web3 URL to EVM Call Message Translation
author: Qi Zhou (@qizhou)
discussions-to: https://ethereum-magicians.org/t/eip-4804-web3-url-to-evm-call-message-translation/8300
type: Standards Track
category: ERC
status: Draft
created: 2022-02-14
requires: N/A
---

## Simple Summary

A translation of an HTTP-style Web3 URL to an EVM call message

## Abstract

The standard translates an HTTP-style Web3 URL like

```
web3://uniswap.eth/
```

to an EVM message such as

```
EVMMessage {
   To: 0xaabbccddee.... // where uniswap.eth's address registered at ENS
   Calldata: 0x
   ...
}
```

## Motivation

Currently, reading data from Web3 generally relies on a translation done by a Web2 proxy to Web3 blockchain. The translation is mostly done by the proxies such as dApp websites/node service provider/etherscan, which are out of the control of users. The standard here aims to provide a simple way for Web2 users to directly access the content of Web3.

## Specification

> The specification here only specifies for the read EVM message. Write message will be defined in the future proposals.

Reviewer: I'd recommend using the Solidity state mutability specifiers (so `view`):

```suggestion
This specification only defines read-only (i.e. Solidity's `view` functions) semantics. State modifying functions may be defined as a future extension.
```

- **chainid** indicates which chain to call the message. If not specified, the protocol will use the same chain as the name service provider, e.g., 1 for eth, and 333 for w3q. If no name service provider is available, the default chainid is 1.
- **returnTypes** tells the format of the returned data. If not specified, the returned message data will be parsed in "(bytes32)" and MIME will be set based on the suffix of the last argument. If **returnTypes** is "()", the returned data will be parsed in raw bytes in JSON.  Otherwise, the returned message will be parsed in the specified "returnTypes" in JSON.  

### Resolve Mode

Once the "To" address and chainid are determined, the protocol will check the resolver mode of contract by calling "resolveMode" method. The protocol currently supports two resolve modes:

#### Manual Mode

The manual mode will not do any interpretation of **path**, and put **path** as calldata of the message directly.

#### Auto Mode

> The auto mode is the default mode of resolver (also applies when the "resolveMode" method is unavailable in the target contract). In the auto mode, if **path** is empty, then the protocol will call the target contract with empty calldata. Otherwise, the calldata of the EVM message will use standard Ethereum contract ABI [Contract ABI Specification — Solidity 0.8.3 documentation](https://docs.soliditylang.org/en/v0.8.3/abi-spec.html), where

Reviewer: Please avoid including external links.

```suggestion
The auto mode is the default mode of resolver (also applies when the "resolveMode" method is unavailable in the target contract). In the auto mode, if **path** is empty, then the protocol will call the target contract with empty calldata. Otherwise, the calldata of the EVM message will use standard Solidity contract ABI, where
```

- **method** is a string of function method be called
- **argument_i** is the ith argument of the method. If **type** is specified, the value will be translated to the corresponding type. The protocol currently supports the basic types such as uint256, bytes32, address, and bytes. If **type** is not specified, then the type will be automatically detected using the following rule in a sequential way:

1. **type** = "uint256", if **value** is numeric; or
2. **type**="bytes32", if **value** is in the form of 0x+32-byte-data hex; or
3. **type**="address", if **value** is in the form of 0x+20-byte-data hex; or
4. **type**="name", if **value** is in the form of **name**.**nsProvider**. In this case, the actual value of the argument will be **address**, which is obtained from **nsProvider**, e.g., eth, w3q, etc.
5. else **type**="bytes"

Note that if **method** does not exist, i.e., **path** is empty or "/", then the contract will be called with empty calldata.

> 

Reviewer: You're missing a rationale section here. I'd go through EIP-1 and make sure everything is there and in order, just in case.

Author: Done.  Please take a look.

> ## Examples

Reviewer: This is not a valid top level section according to [EIP-1](https://eips.ethereum.org/EIPS/eip-1).

Author: Changed to ### Examples

#### Example 1

```
web3://ensdomains.eth:4
```

The protocol will find the address of **ensdomains.eth** from ENS in chainid 4 (Rinkeby), and then call the address with "From" = "0x..." and "Calldata" = "0x" with chainid = 4.

#### Example 2

```
web3://0x9e081Df45E0D167636DB9C61C7ce719A58d82E3b:4
```

The protocol will call the address with "From" = "0x9e081Df45E0D167636DB9C61C7ce719A58d82E3b" and "Calldata" = "0x" with chainid = 4.

#### Example 3

```
web3://cyberbrokers-meta.eth:1/renderBroker/9999
```

The protocol will find the address of **cyberbrokers-meta.eth** from ENS, and then call the address with "From" = "0x..." and "Calldata" = "0x" + `keccak("view(uint256)")[0:4] + abi.encode(uint256(9999))` with chainid = 1.

#### Exmaple 4

```
web3://home.w3q:3334/
```

> The protocol will find the address of **home.w3q** from W3NS, and then call the address with "From" = "0x..." and "Calldata" = "0x" with chainid = 3334.

Reviewer: I think this EIP is going to need a list of `nsProvider` suffixes, and how to resolve them.

Author: A section about how to solve is added.  Please take a look.

#### Resolving Address from W3NS

Given **contractName**, the protocol will find the address of the contract using the following steps:
1. Find the Web handler record on W3NS resolver.  Return error if the record is an invalid ETH address.
2. If the Web handler record does not exist, the protocol will use ETH address of the name.

Given **name** in arguments, the protocol will find the address of the name using its ETH address.

### Examples

#### Example 1

```
web3://ensdomains.eth:4
```

The protocol will find the address of **ensdomains.eth** from ENS in chainid 4 (Rinkeby), and then call the address with "From" = "0x..." and "Calldata" = "0x" with chainid = 4.

#### Example 2

```
web3://0x9e081Df45E0D167636DB9C61C7ce719A58d82E3b:4
```

The protocol will call the address with "From" = "0x9e081Df45E0D167636DB9C61C7ce719A58d82E3b" and "Calldata" = "0x" with chainid = 4.

#### Example 3

```
web3://cyberbrokers-meta.eth:1/renderBroker/9999
```

The protocol will find the address of **cyberbrokers-meta.eth** from ENS, and then call the address with "From" = "0x..." and "Calldata" = "0x" + `keccak("view(uint256)")[0:4] + abi.encode(uint256(9999))` with chainid = 1.

> #### Exmaple 4

Reviewer: ```suggestion
#### Example 4
```

## Rationale

The purpose of the proposal is to add a decentralized presentation layer for Ethereum.  With the layer, we are able to render any web content (including HTML/CSS/JPG/PNG/SVG, etc) on-chain using human-readable URLs, and thus EVM can be served as decentralized Backend.  The design of the standard is based on the following principles:

- **Human-readable**.  The Web3 URL should be easily recognized by human similar to Web2 URL (http://).  As a result, we support names from name services to replace address for better readability.  In addition, instead of using calldata in hex, we use human-readable method + arguments and translate them to calldata for better readability.

- **Maximum-Compatible with HTTP-URL standard**.  The Web3 URL should be compatible with HTTP-URL standard including relative pathing, query, fragment, etc so that the support of existing HTTP-URL (e.g., by browser) can be easily extended to Web3 URL with minimal modification.  This also means that existing Web2 users can easily migrate to Web3 with minimal extra knowledge of this standard.

> - **Simple**.  Instead of providing explict types in arguments, we use a "maximum likelihood" principle of auto-detecting the types of the arguments such as address, bytes32, and uint256.  This could greatly minimize the length of URL, while avoiding confusion.  In addition, explict types are also supported to clear the confusion if necessary.

Reviewer: ```suggestion
- **Simple**.  Instead of providing explicit types in arguments, we use a "maximum likelihood" principle of auto-detecting the types of the arguments such as address, bytes32, and uint256.  This could greatly minimize the length of URL, while avoiding confusion.  In addition, explicit types are also supported to clear the confusion if necessary.
```

- **Flexible**.  The contract is able to override the encoding rule so that the contract has fine-control of understanding the actual Web resources that the users want to locate.

> 

Reviewer: The "Security Considerations" section is mandatory, though you can probably just do:

```suggestion
## Security Considerations

No security considerations were found.
```

---
eip: 4804
title: Web3 URL to EVM Call Message Translation
description: A translation of an HTTP-style Web3 URL to an EVM call message
author: Qi Zhou (@qizhou), Chao Pi (@pichaoqkc), Sam Wilson (@SamWilsn)
discussions-to: https://ethereum-magicians.org/t/eip-4804-web3-url-to-evm-call-message-translation/8300
status: Draft
type: Standards Track
category: ERC
created: 2022-02-14
---

## Abstract

> This standard translates an RFC 2396 URI like `web3://uniswap.eth/` to an EVM message such as:

Reviewer 2: I'm not sure I understand what you mean by "EVM message" ? This isn't very common terminology - I think it would be better to say something more along the lines of "it's a standard way to represent an `eth_call`" (h/t @SamWilsn ).

Author: Perhaps `EVM message` is a bit old term as I borrowed from pyethereum.  It represents a call to an address with data such as `calldata`, `gasPrice`, etc.   The following is the code snippet copied from pyethereum repo.

![image](https://user-images.githubusercontent.com/2541286/166499989-f48cad6a-9cb7-4859-9076-b6d6f3d63981.png)

https://github.com/ethereum/pyethereum/blob/2f59abb10715049196aeef062931f65495a6bf8d/ethereum/fastvm.py#L63

Meanwhile, looks like in `geth`, it does have explicit `Message` term and just call a `Contract` instead. https://github.com/ethereum/go-ethereum/blob/53304ff6c7b6afdff0f6c338e5480b8112080e2c/core/vm/contract.go#L45

I am open to see what is the better term to describe here.

Author: Just re-examine Geth and found the Call Message is defined as follows.

https://github.com/ethereum/go-ethereum/blob/0169d579d0eed4f6366697985a7b0f0b99402783/interfaces.go#L134

### Resolving Address from Name Service Providers

The protocol currently supports two name services:
- Ethereum name service with suffix `.eth`; and
- W3 name service with suffix `.w3q`.

#### Resolving Address from ENS
Given **contractName**, the protocol will find the address of the contract using the following steps:
1. Find the `w3q` text record on ENS resolver.  Return error if the record is an invalid ETH address.
2. If the `w3q` text record does not exist, the protocol will use ETH address of the name.

Given **name** in arguments, the protocol will find the address of the name using its ETH address.

#### Resolving Address from W3NS

Given **contractName**, the protocol will find the address of the contract using the following steps:
1. Find the Web handler record on W3NS resolver.  Return error if the record is an invalid ETH address.
2. If the Web handler record does not exist, the protocol will use ETH address of the name.

Given **name** in arguments, the protocol will find the address of the name using its ETH address.

### Examples

#### Example 1

```
web3://ensdomains.eth:4
```

> The protocol will find the address of **ensdomains.eth** from ENS in chainid 4 (Rinkeby), and then call the address with "From" = "0x..." and "Calldata" = "0x" with chainid = 4.

Reviewer 2: What is the from address in this example?

#### Example 2

```
web3://0x9e081Df45E0D167636DB9C61C7ce719A58d82E3b:4
```

> The protocol will call the address with "From" = "0x9e081Df45E0D167636DB9C61C7ce719A58d82E3b" and "Calldata" = "0x" with chainid = 4.

Reviewer 2: This From should be empty, no?

- **Maximum-Compatible with HTTP-URL standard**.  The Web3 URL should be compatible with HTTP-URL standard including relative pathing, query, fragment, etc so that the support of existing HTTP-URL (e.g., by browser) can be easily extended to Web3 URL with minimal modification.  This also means that existing Web2 users can easily migrate to Web3 with minimal extra knowledge of this standard.

- **Simple**.  Instead of providing explicit types in arguments, we use a "maximum likelihood" principle of auto-detecting the types of the arguments such as address, bytes32, and uint256.  This could greatly minimize the length of URL, while avoiding confusion.  In addition, explicit types are also supported to clear the confusion if necessary.

- **Flexible**.  The contract is able to override the encoding rule so that the contract has fine-control of understanding the actual Web resources that the users want to locate.

## Security Considerations

No security considerations were found.

> ## Copyright
> 

Reviewer 3: ```suggestion

## Copyright
```

Reviewer 3: We are looking into automated checking of this section, which requires no newline after `## Copyright`.  Also, you are missing a newline before it.

> Copyright and related rights waived via [CC0 1](https://creativecommons.org/publicdomain/zero/1.0/).

Reviewer 4: ```suggestion
Copyright and related rights waived via [CC0](../LICENSE).
```

- **Maximum-Compatible with HTTP-URL standard**.  The Web3 URL should be compatible with HTTP-URL standard including relative pathing, query, fragment, etc so that the support of existing HTTP-URL (e.g., by browser) can be easily extended to Web3 URL with minimal modification.  This also means that existing Web2 users can easily migrate to Web3 with minimal extra knowledge of this standard.

- **Simple**.  Instead of providing explict types in arguments, we use a "maximum likelihood" principle of auto-detecting the types of the arguments such as address, bytes32, and uint256.  This could greatly minimize the length of URL, while avoiding confusion.  In addition, explict types are also supported to clear the confusion if necessary.

- **Flexible**.  The contract is able to override the encoding rule so that the contract has fine-control of understanding the actual Web resources that the users want to locate.

> ### Examples

Reviewer: This whole section could actually go in the specification section, instead of rationale.

Author: Done.  Thanks for the comment.

```

The protocol will find the address of **wusdt.eth** and then call the method "balanceOf(address)" of the address. The returned data will be parsed as raw bytes like `["0x000000000000000000000000000000000000000000000000000009184e72a000"]`.

### Resolving Address from Name Service Providers

The protocol currently supports two name services:
- Ethereum name service with suffix `.eth`; and
- W3 name service with suffix `.w3q`.

#### Resolving Address from ENS
Given **contractName**, the protocol will find the address of the contract using the following steps:
1. Find the `w3q` text record on ENS resolver.  Return error if the record is an invalid ETH address.
2. If the `w3q` text record does not exist, the protocol will use ETH address of the name.

Given **name** in arguments, the protocol will find the address of the name using its ETH address.

#### Resolving Address from W3NS

Given **contractName**, the protocol will find the address of the contract using the following steps:
1. Find the Web handler record on W3NS resolver.  Return error if the record is an invalid ETH address.
2. If the Web handler record does not exist, the protocol will use ETH address of the name.

> Given **name** in arguments, the protocol will find the address of the name using its ETH address.

Reviewer: These should be in `Specification` section.

Author: Done.  Thanks for the comment.

### Reference Implementation

The above Web3 links are supported by Web3Q gateway by browsing https://web3q.io/url_without_web3_schema, e.g.,
- https://web3q.io/ensdomains.eth:4/.  Decentralized dWeb hosting of ENS homepage on Rinkeby testnet.

> - https://web3q.io/cyberbrokers-meta.eth:1/renderBroker/2999.  Decentralized NFT metadata rendering on Mainnet.

Reviewer: A reference implementation should include the actual source code here (either inline, or in the `assets/eip-4804` directory.) Without the source code, it isn't much of a reference for other people :P

That said, the reference implementation is optional.

Author: Removed.  Thanks for the comment.
